Fix IPv6 address selection for deprecated addresses

jukkar opened the pull request:
Always prefer preferred IPv6 address over deprecated one regardless of prefix length. This works now same way as in Linux.

jukkar committed:
net: if: Prefer always preferred IPv6 address

Always prefer preferred IPv6 address over deprecated one
regardless of prefix length. This works now same way as in
Linux.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: ipv6: Fix deprecated address selection

Make sure the IPv6 address selection works as specified.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

Damian-Nordic commented on subsys/net/ip/net_if.c:
```diff
@@ -3182,14 +3182,15 @@ static struct in6_addr *net_if_ipv6_get_best_match(struct net_if *iface,
 
-			if (len == *best_so_far &&
-			    ipv6->unicast[i].addr_state == NET_ADDR_DEPRECATED &&
+			if (ipv6->unicast[i].addr_state == NET_ADDR_DEPRECATED &&
```
This addresses the case where deprecated address comes after preferred one, but there's another scenario in which the preferred address may fail the `len >= *best_so_far` condition and hence it may not be selected over the deprecated address, right?

jukkar replied:
There is a check
```
	if (len >= *best_so_far) {
```
few lines above. It should select the best prefix, we then skip the deprecated address if the other one is preferred one. I tried to construct the tests so that they verify this. Please clarify more what you mean if I missed something.

Damian-Nordic replied:
I think the order of processing addresses shouldn't matter.

In case the address array contains:
[0] Preferred address for which `get_diff_ipv6` returns 32
[1] Deprecated address for which `get_diff_ipv6` returns 64

the code will correctly choose [0] despite [1] being a better match prefix-wise.

However if the order of elements is opposite:
[0] Deprecated address for which `get_diff_ipv6` returns 64
[1] Preferred address for which `get_diff_ipv6` returns 32

the code will choose [0] because [1] will be discarded early due to `len >= *best_so_far` not being satisfied (`32 >= 64`). I think the code should define some kind of `is_address_better` predicate and this predicate should exit early whenever one address is preferred and another one is deprecated.

rlubos replied:
I think @Damian-Nordic is right, this will fail if the deprecated address is checked first, for example:
1. We first check deprecated address, with 64-bit prefix match. `best_so_far` is set to 64
2. In the second iteration we check preferred address with 40-bit prefix match. The preferred address should take preference, but since the prefix match is worse than `best_so_far`, we won't even check if the current address is deprecated.

I think we need to check `addr_state` for deprecated address before we do the prefix-match check.

jukkar replied:
Indeed, good point. I will rework the code.

jukkar committed:
net: if: Add debug prints for IPv6 prefix selection

Useful to see what prefix is being selected for a given
IPv6 address if debugging is enabled.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

Damian-Nordic commented on subsys/net/ip/net_if.c:
```diff
@@ -2864,6 +2864,16 @@ struct net_if_ipv6_prefix *net_if_ipv6_prefix_get(struct net_if *iface,
 out:
 	net_if_unlock(iface);
 
+	if (prefix != NULL) {
```
You may want to do this before `net_if_unlock` to make sure the prefix is not changed in the middle of printing it.

Damian-Nordic commented on subsys/net/ip/net_if.c:
```diff
@@ -3173,6 +3183,18 @@ static struct in6_addr *net_if_ipv6_get_best_match(struct net_if *iface,
 			len = prefix_len;
 		}
 
+		if (ipv6->unicast[i].addr_state == NET_ADDR_DEPRECATED &&
```
Just moving it here is not enough. What you need is turning
```
if (len >= *best_so_far) {
```
into sth like
```
if (len >= *best_so_far || (ipv6->unicast[i].addr_state == NET_ADDR_PREFERRED && addr_state == NET_ADDR_DEPRECATED) {
```
right?

jukkar replied:
If we always prefer preferred address instead of deprecated, then we do not really need to check the prefix length as it should not matter here. Note that the `addr_state` variable stores the state of the current best address.

Damian-Nordic replied:
> If we always prefer preferred address instead of deprecated, then we do not really need to check the prefix length as it should not matter here

Yes, and this is what we need - in case A is preferred and B is deprecated we don't care whether CommonPrefixLength(A, destination) is longer than CommonPrefixLength(B, destination).

jukkar replied:
> The common prefix length should only matter when comparing deprecated to deprecated or preferred to preferred.

That should happen with the current code in the PR, the tests should verify it.

jukkar replied:
Output from running the tests
```
START - test_ipv6_addresses
...
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_prefix_add: (test_ipv6_addresses): [0] interface 0x806e9e4 prefix fc00:aaaa::d1d2:d3d4/96 added
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_prefix_add: (test_ipv6_addresses): [1] interface 0x806e9e4 prefix fc00:aaaa::d1d2:d302/64 added
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_addr_add: (test_ipv6_addresses): [0] interface 1 (0x806e9e4) address fc00:aaaa::d1d2:d3d4 type AUTO added
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_prefix_get: (test_ipv6_addresses): Found prefix fc00:aaaa::d1d2:d302/64 for fc00:aaaa::d1d2:d302
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): [0] Checking fc00:aaaa::d1d2:d3d4 (preferred) dst fc00:aaaa::d1d2:d302/64
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): Selected fc00:aaaa::d1d2:d3d4 (preferred) dst fc00:aaaa::d1d2:d302/64
448: Selected IPv6 address fc00:aaaa::d1d2:d3d4 state preferred, iface (nil)
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_prefix_get: (test_ipv6_addresses): Found prefix fc00:aaaa::d1d2:d302/64 for fc00:aaaa::d1d2:f303
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): [0] Checking fc00:aaaa::d1d2:d3d4 (deprecated) dst fc00:aaaa::d1d2:f303/64
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): Selected fc00:aaaa::d1d2:d3d4 (deprecated) dst fc00:aaaa::d1d2:f303/64
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_addr_add: (test_ipv6_addresses): [1] interface 1 (0x806e9e4) address fc00:aaaa::d1d2:f303 type AUTO added
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_prefix_get: (test_ipv6_addresses): Found prefix fc00:aaaa::d1d2:d302/64 for fc00:aaaa::d1d2:d302
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): [0] Checking fc00:aaaa::d1d2:d3d4 (preferred) dst fc00:aaaa::d1d2:d302/64
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): [1] Checking fc00:aaaa::d1d2:f303 (preferred) dst fc00:aaaa::d1d2:d302/64
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): Selected fc00:aaaa::d1d2:f303 (preferred) dst fc00:aaaa::d1d2:d302/64
480: Selected IPv6 address fc00:aaaa::d1d2:f303 state preferred, iface (nil)
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_prefix_get: (test_ipv6_addresses): Found prefix fc00:aaaa::d1d2:d302/64 for fc00:aaaa::d1d2:d302
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): [0] Checking fc00:aaaa::d1d2:d3d4 (preferred) dst fc00:aaaa::d1d2:d302/64
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): skipping deprecated address fc00:aaaa::d1d2:f303
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): Selected fc00:aaaa::d1d2:d3d4 (preferred) dst fc00:aaaa::d1d2:d302/64
496: Selected IPv6 address fc00:aaaa::d1d2:d3d4 state preferred, iface (nil)
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_prefix_get: (test_ipv6_addresses): Found prefix fc00:aaaa::d1d2:d302/64 for fc00:aaaa::d1d2:d302
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): [0] Checking fc00:aaaa::d1d2:d3d4 (deprecated) dst fc00:aaaa::d1d2:d302/64
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): [1] Checking fc00:aaaa::d1d2:f303 (deprecated) dst fc00:aaaa::d1d2:d302/64
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): Selected fc00:aaaa::d1d2:f303 (deprecated) dst fc00:aaaa::d1d2:d302/64
510: Selected IPv6 address fc00:aaaa::d1d2:f303 state deprecated, iface (nil)
```
Above were two deprecated addresses and we select the best from those two.
```
cont...
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_addr_add: (test_ipv6_addresses): [2] interface 1 (0x806e9e4) address fc00:aaaa::d1d2:f304 type AUTO added
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_prefix_get: (test_ipv6_addresses): Found prefix fc00:aaaa::d1d2:d302/64 for fc00:aaaa::d1d2:d3d5
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): [0] Checking fc00:aaaa::d1d2:d3d4 (deprecated) dst fc00:aaaa::d1d2:d3d5/64
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): [1] Checking fc00:aaaa::d1d2:f303 (preferred) dst fc00:aaaa::d1d2:d3d5/64
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): skipping deprecated address fc00:aaaa::d1d2:f304
[00:00:00.000,000] <dbg> net_if: net_if_ipv6_get_best_match: (test_ipv6_addresses): Selected fc00:aaaa::d1d2:f303 (preferred) dst fc00:aaaa::d1d2:d3d5/64
531: Selected IPv6 address fc00:aaaa::d1d2:f303 state preferred, iface (nil)
[00:00:00.000,000] <dbg> net_if: net_if_addr_unref_debug: (test_ipv6_addresses): [1] ifaddr fc00:aaaa::d1d2:d3d4 state 2 ref 0 (net_if_ipv6_addr_rm():2107)
[00:00:00.000,000] <dbg> net_if: net_if_addr_unref_debug: (test_ipv6_addresses): [1] ifaddr fc00:aaaa::d1d2:f303 state 1 ref 0 (net_if_ipv6_addr_rm():2107)
[00:00:00.000,000] <dbg> net_if: net_if_addr_unref_debug: (test_ipv6_addresses): [1] ifaddr fc00:aaaa::d1d2:f304 state 2 ref 0 (net_if_ipv6_addr_rm():2107)
 PASS - test_ipv6_addresses in 0.000 seconds
===================================================================
```

jukkar commented:
* Updated according to comments

Damian-Nordic reviewed (approved):
Thanks!

jukkar commented:
@pdgendt some help would be needed in review, please take a look if you have time, thanks!
